actors/migration/nv3/miner.go

```diff
@@ -1,30 +1,472 @@
+	}
+
+	dlInfo := st.DeadlineInfo(epoch)
+	crontMM, err := adt.AsMultimap(store, powSt.CronEventQueue)
+	if err != nil {
+		return err
+	}
+	err = crontMM.Add(abi.IntKey(int64(dlInfo.Last())), &event)
+	if err != nil {
+		return err
+	}
+
+	powSt.CronEventQueue, err = crontMM.Root()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *minerMigrator) updateClaim(ctx context.Context, store adt.Store, st *miner.State, powSt *power.State, a addr.Address) error {
+	deadlines, err := st.LoadDeadlines(store)
+	if err != nil {
+		return err
+	}
+
+	activePower := miner.NewPowerPairZero()
+	err = deadlines.ForEach(store, func(dlIdx uint64, dl *miner.Deadline) error {
+		partitions, err := dl.PartitionsArray(store)
+		if err != nil {
+			return err
+		}
+
+		var part miner.Partition
+		return partitions.ForEach(&part, func(pIdx int64) error {
+			activePower = activePower.Add(part.LivePower.Sub(part.FaultyPower))
+			return nil
+		})
+	})
+	if err != nil {
+		return err
+	}
+
+	claims, err := adt.AsMap(store, powSt.Claims)
+	if err != nil {
+		return err
+	}
+
+	var existing power.Claim
+	powerDelta := activePower
+	found, err := claims.Get(abi.AddrKey(a), &existing)
+	if err != nil {
+		return err
+	}
+	if found {
+		powerDelta = powerDelta.Sub(miner.NewPowerPair(existing.RawBytePower, existing.QualityAdjPower))
+	}
+	if err := powSt.AddToClaim(store, a, powerDelta.Raw, powerDelta.QA); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type expirationQueueStats struct {
+	// total of all active power in the expiration queue
+	totalActivePower miner.PowerPair
+	// total of all faulty power in the expiration queue
+	totalFaultyPower miner.PowerPair
+}
+
+// Updates the expiration queue by correcting any duplicate entries and their fallout.
+// If no changes need to be made cid.Undef will be returned.
+// Returns the new root of the expiration queue
+func (m *minerMigrator) correctExpirationQueue(exq miner.ExpirationQueue, sectors miner.Sectors,
+	allTerminated bitfield.BitField, allFaults bitfield.BitField, sectorSize abi.SectorSize,
+) (cid.Cid, expirationQueueStats, error) {
+	// processed expired sectors includes all terminated and all sectors seen in earlier expiration sets
+	processedExpiredSectors := allTerminated
+	expirationSetPowerSuspect := false
+
+	var exs miner.ExpirationSet
+
+	// Check for faults that need to be erased.
+	// Erased faults will be removed from bitfields and the power will be recomputed
+	// in the subsequent loop.
+	err := exq.ForEach(&exs, func(epoch int64) error { //nolint:nomaprange
+		// Detect sectors that are present in this expiration set as "early", but that
+		// have already terminated or duplicate a prior entry in the queue, and thus will
+		// be terminated before this entry is processed. The sector was rescheduled here
+		// upon fault, but the entry is stale and should not exist.
+		modified := false
+		earlyDuplicates, err := bitfield.IntersectBitField(exs.EarlySectors, processedExpiredSectors)
```

Comment:
Wouldn't `bitfield.SubtractBitField(exs.EarlySectors, liveSectors)` be safer? That would handle completely removed/compacted out sectors.

Note: this code should work as-is assuming no compaction, which is probably safe.

Reply:
Same below.